[RFC] Codify CI policy

ppannuto opened the pull request:
### Pull Request Overview

This takes a first cut at writing down a structure and policy for CI.

### Testing Strategy

N/A.

### TODO or Help Wanted

Feedback!

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make format`.
- [ ] Fixed errors surfaced by `make clippy`.

---

[Rendered](https://github.com/tock/tock/blob/ci-policy/doc/CodeReview.md#3-continuous-integration)

alistair23 commented on doc/CodeReview.md:
```diff
@@ -77,7 +78,87 @@ possible responses:
+locally, although in certain cases this may require installing supporting
+tooling.
+
+##### `ci-setup-*`
```
Will `make ci` run `ci-setup-*` and then `ci-job-*`

ppannuto replied:
Effectively, thought the setup would be more of:
```make
ci-job-format-check: ci-setup-rustfmt
     run rustfmt

ci-setup-rustfmt:
     install rustfmt

ci:  ci-job-format-check
```

alistair23 commented on doc/CodeReview.md:
```diff
@@ -77,7 +78,87 @@ possible responses:
+This is the rule that should be run locally before submitting PRs.
+It runs the quicker jobs that catch the majority of small errors.
+
+
```
Double space here

lschuermann commented on doc/CodeReview.md:
```diff
@@ -77,7 +85,86 @@ possible responses:
+Setup rules may vary between runner and local environments, as they may perform
+automatic and possibly invasive (e.g. apt install) operations on runners.
+
+When run locally, setup targets **may** prompt users during install events.
```
Mhm. This implies that a `make ci` could potentially persistently and arbitrarily change my system by doing system-wide installs and or configuration of tools, right? I think that this _may_ be helpful for some people, however there are probably a lot of people (including me) who do not want this.

On a local (non CI-infrastructure) build, I don't expect any persistent changes of my system outside of the tock working directory, or _at least_ a polite question whether the program is allowed to do so. I therefore propose changing this to a `setup targets **must** prompt users prior to system-wide persistent changes`.

alistair23 replied:
Agreed. I think in general we will avoid installing anything (as it's annoying and hard to do anyway).

ppannuto commented:
Updated the setup language.

Changed `make ci` to `make prepush` to indicate its intent. @hudson-ayers that target won't be the same as what's currently defined as `make ci` [the currently defined `make ci` would be the `make ci-all` rule in this proposal]

hudson-ayers commented on doc/CodeReview.md:
```diff
@@ -77,7 +85,95 @@ possible responses:
+This is a meta-target that runs what Tock considers the "standard developer CI".
+This is the rule that should be run locally before submitting PRs.
+It runs the quicker jobs that catch the majority of small errors.
+Developers are encouraged to consider wiring this to the
```
```suggestion
Developers are encouraged to consider writing this to the
```

ppannuto replied:
I think I meant wiring?

hudson-ayers replied:
Oh, I see, I am dumb, sorry

brghena commented:
bors r+
